[FIX] account: prevent wrong floating representation of currency

yosa-odoo opened the pull request:
Steps to reproduce:
- In Journals/Bills/Advanced activate "Lock post entries with hash"
- Create a purchase tax of 17%
- Create a bill with a product of a cost of 30 and apply the "17%" tax
- Post it
- Export the data inalterability check report

Issue:
"Corrupted data"

Cause:
In python we have a reprentation issue
```
>>> 30*0.17
5.1000000000000005
```
We define the hash string during the `_compute_string_to_hash` at the
move creation. The issue is, at that time, the move_line tax debit
is not rounded.

Therefore, when we print the report, we take the `move.line_id.
debit` from the db which is rounded and equal to '5.10' and compare
it to '5.1000000000000005' which gives a different hash

Solution:
Implementing a V3 version that uses `repr` for monetary fields.

opw-3072693

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/112016).

william-andre reviewed (changes requested):
We can't change the hash function in the middle of an existing chain

william-andre reviewed (approved):
Nevermind, you can in this case
@robodoo r+

There is no way to fix the customer's issue though, since the value used to build the chain has been lost (but it can be explained to an auditor if needed)

william-andre commented:
```
>>> move.line_ids[0].credit
2.5
>>> float_repr(move.line_ids[0].credit, 2)
'2.50'
```
@robodoo r-

yosa-odoo commented:
```
>>> move2.line_ids[0].credit
5.1000000000000005
>>> field_value = move2.line_ids[0].credit
>>> len(str(field_value).split('.')[-1]) > move2.currency_id.decimal_places
True
>>> move.line_ids[0].credit
2.5
>>> field_value = move.line_ids[0].credit
>>> len(str(field_value).split('.')[-1]) > move.currency_id.decimal_places
False
```

yosa-odoo commented:
waiting for https://github.com/odoo/odoo/pull/112320 to be merged

rigr-odoo commented:
@yosa-odoo 
https://github.com/odoo/odoo/pull/112320 is finally merged so you can implement v3 as we have already talked about :D

william-andre commented on addons/account/tests/test_account_inalterable_hash.py:
```diff
@@ -188,3 +188,81 @@ def test_account_move_hash_versioning_v1_to_v2(self):
         integrity_check = moves.company_id._check_hash_integrity()['results'][0]
         self.assertEqual(integrity_check['msg_cover'], f'Corrupted data on journal entry with id {moves_v1_bis[0].id}.')
+
+    def test_account_move_hash_v2_floating_point(self):
```
It's nice to have this test and see that it is green while merging, but it is not achieving anything for the CI later on

william-andre replied:
The next test does that.
This test makes sure that it was failing before the fix, which we don't really care about. We just don't want it to fix in the future.

william-andre commented on addons/account/tests/test_account_inalterable_hash.py:
```diff
@@ -188,3 +188,81 @@ def test_account_move_hash_versioning_v1_to_v2(self):
+        )
+        moves_v3.with_context(hash_version=3).action_post()
+        fields_v3 = moves_v3._get_integrity_hash_fields()
+        self.assertEqual(fields_v2, fields_v3)
```
Not sure why we need to check this.
Note that `moves_v2` and `moves_v3` have the same context (without `hash_version` set, therefore there is no reason for `_get_integrity_hash_fields` to return something different-

rigr-odoo replied:
> Not sure why we need to check this.

For the same reason there was a test v1 to v2, i.e. check that the integrity report goes through all possible versions without errors

> Note that moves_v2 and moves_v3 have the same context (without hash_version set, therefore there is no reason for _get_integrity_hash_fields to return something different-

Indeed, perhaps @yosa-odoo could test `_getattrstring` instead

william-andre replied:
> For the same reason there was a test v1 to v2

I mean the specific assert I'm commenting on. We don't really care about the return result of private functions, we care about user flows.

> Indeed, perhaps @yosa-odoo could test `_getattrstring` instead

? :suspect:

yosa-odoo replied:
Not sure to understand the `_getattrstring` but we can indeed only check that the report is flawless with use of version 2 and version 3

william-andre reviewed (approved):
@robodoo r+
